stdlib/prelude.an

```diff
@@ -674,23 +674,24 @@ next_line (f: InFile) : string =
     array_insert (@cstr) len '\0'
     string (@cstr) len
 
-
-
-/// Print a message then return user input
-input msg : string =
-    printne msg
-
+// Read a single line from stdin
+read_line () : string =
     buffer, len = loop (buffer = null ()) (len = 0) ->
         c = getchar ()
         buffer2 = if len % 32 == 0
             then realloc buffer (len + 32)
             else buffer
 
-        buffer2#len := c
+        array_insert buffer2 len c
 
-        if c != '\n'
+        if c != '\n' and (transmute c : i8) >= 0
```

Comment:
I believe EOF should actually be 255, where 0 is the NUL character (there is also a literal for it: `'\0'`). `next_line` checks for 255 as well for example. You can verify its working by calling the read_line function and hitting Ctrl+D after typing some text to issue the EOF character normally. I tried changing it to this locally and it does seem like I need to hit Ctrl+D twice for some reason.

Reply:
Oh, I see what you mean. EOF is denoted by `-1` when calling `getchar()` in C, since it [returns an int](https://stackoverflow.com/a/19716188/2161072) rather than an unsigned char. But in Ante's prelude, we declare it as `unit -> char` so it's downcasted to 8 bits and turns into 255 due to the 2's complement representation.

Normal input can contain ASCII value 255 without intending to end the input, so... maybe we should change `getchar` to match C semantics?

Reply:
@jfecher how about the latest change, declaring `getchar` to return int, and adding `read_char` to return a `Maybe char` taking EOF into account? I _think_ the logic in `read_line` is still correct (although I managed to produce a segfault when returning the old buffer by accident, should be good now).

Reply:
Sounds good, though it doesn't look like the latest change you mentioned has been pushed

Reply:
Wow, I really do need sleep :laughing: Pushed now.

Reply:
Hmm, I wonder if `read_line` should also return `Maybe string`. Otherwise you probably have to keep manually checking for EOF (some other way...).

Reply:
Oops... I changed it to do that, got it working in one program, but failing in another for some reason I don't understand:

```
thread 'main' panicked at '- inst0: The bitcast argument v0 doesn't fit in a type of 8 bits
', src/cranelift_backend/context.rs:194:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_display
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:72:5
   3: ante::cranelift_backend::context::Context::codegen_function_body
             at /home/omf/code/ante/compiler/src/cranelift_backend/context.rs:194:13
   4: ante::cranelift_backend::context::Context::codegen_all
             at /home/omf/code/ante/compiler/src/cranelift_backend/context.rs:152:13
   5: ante::cranelift_backend::run
             at /home/omf/code/ante/compiler/src/cranelift_backend/mod.rs:23:5
   6: ante::compile
             at /home/omf/code/ante/compiler/src/main.rs:171:9
   7: ante::main
             at /home/omf/code/ante/compiler/src/main.rs:100:9
   8: core::ops::function::FnOnce::call_once
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
(With the following one-line test program: `s = read_line ()`)

Reply:
Interestingly, it compiles just fine if I add `-O3` :thinking: